src/App/Fossa/Reachability/Upload.hs

```diff
@@ -0,0 +1,154 @@
+import Srclib.Types (
+  Locator,
+  SourceUnit (..),
+  SourceUnitBuild (..),
+  SourceUnitDependency (..),
+ )
+import Types (DiscoveredProjectType (..), GraphBreadth (..))
+
+analyzeForReachability ::
+  ( Has Logger sig m
+  , Has ReadFS sig m
+  , Has Diagnostics sig m
+  , Has Exec sig m
+  , Has (Lift IO) sig m
+  , Has Debug sig m
+  ) =>
+  AnalysisScanResult ->
+  m [SourceUnitReachability]
+analyzeForReachability analysisResult = context "reachability" $ do
+  let analyzerResult = analyzersScanResult analysisResult
+  units <- catMaybes <$> (traverse callGraphOf analyzerResult)
+  debugMetadata reachabilityRawJson units
+  pure units
+
+upload ::
+  ( Has FossaApiClient sig m
+  , Has Debug sig m
+  ) =>
+  ProjectRevision ->
+  ProjectMetadata ->
+  [SourceUnitReachability] ->
+  m ()
+upload revision metadata units = do
+  units' <- traverse uploadReachability units
+  debugMetadata reachabilityEndpointJson units'
+  uploadBuildForReachability revision metadata units'
+
+uploadReachability ::
+  ( Has FossaApiClient sig m
+  ) =>
+  SourceUnitReachability ->
+  m SourceUnitReachability
+uploadReachability unit = case callGraphAnalysis unit of
+  NoCallGraphAnalysis -> pure unit
+  JarAnalysis [] -> pure unit
+  JarAnalysis someJars -> do
+    updatedJars <- traverse uploadJarAnalysis someJars
+    pure $ unit{callGraphAnalysis = JarAnalysis updatedJars}
+
+uploadJarAnalysis ::
+  ( Has FossaApiClient sig m
+  ) =>
+  ParsedJar ->
+  m ParsedJar
+uploadJarAnalysis jar = case parsedJarContent jar of
+  ContentStoreKey _ -> pure jar
+  ContentRaw bs -> do
+    key <- uploadContentForReachability bs
+    pure $ jar{parsedJarContent = ContentStoreKey key}
+
+callGraphOf ::
+  ( Has Logger sig m
+  , Has ReadFS sig m
+  , Has Diagnostics sig m
+  , Has Exec sig m
+  , Has (Lift IO) sig m
+  ) =>
+  DiscoveredProjectScan ->
+  m (Maybe SourceUnitReachability)
+callGraphOf (Scanned dpi (Success _ projectResult)) = do
+  let srcUnit = projectToSourceUnit False projectResult
+  let dependencies = dependenciesOf srcUnit
+  let displayId = sourceUnitType srcUnit <> "@" <> sourceUnitManifest srcUnit
+  let unit =
+        SourceUnitReachability
+          { srcUnitType = sourceUnitType srcUnit
+          , srcUnitManifest = sourceUnitManifest srcUnit
+          , srcUnitName = sourceUnitName srcUnit
+          , srcUnitOriginPaths = sourceUnitOriginPaths srcUnit
+          , srcUnitDependencies = dependencies
+          , callGraphAnalysis = NoCallGraphAnalysis
+          }
+  case (projectResultGraphBreadth projectResult, dpiProjectType dpi) of
+    -- if we do not have complete graph, i.e.. missing transitive dependencies
+    -- it is impossible to perform reachability, as we may not have all symbols
+    -- used in the application to perform accurate analysis
+    (Partial, _) -> do
+      logInfo . pretty $ "We do not support reachability analysis, with partial dependencies graph (skipping: " <> displayId <> ")"
```

Comment:
[optional] consider "FOSSA CLI does not ..." rather than "We do not ..." here and in the other log message